modules/roles/server/controllers/roles.controller.js

```diff
@@ -31,191 +47,255 @@ function roleModule(logger, userModule) {
 
 
   /*
-  * Roles can be added to the role tree at any level other than root.
-  * When a role is added, the tree is traversed, and for each 'parent'
-  * the role has, any users with role equal to that parent role will 
-  * get the new role in their subroles array.
-  * 
-  * @param {req}  The Express HTTP request with the role in the body
-  * @param {res} The Express HTTP response object
-  * @param {next} The Express HTTP next function
-  *
-  * @return {bool} whether or not the role was added
-  */
-  function addRole(req, res, next) 
+   * Roles can be added to the role tree at any level other than root.
+   * When a role is added, the tree is traversed, and for each 'parent'
+   * the role has, any users with role equal to that parent role will 
+   * get the new role in their subroles array.
+   * 
+   * @param {req}  The Express HTTP request with the role in the body
+   * @param {res} The Express HTTP response object
+   * @param {next} The Express HTTP next function
+   *
+   * @return {bool} whether or not the role was added
+   */
+  function addRole(req, res, next)
   {
-    var role = req.body;
-    //make sure the role is valid
-    if(role._id != null)
+    //validation check
+    if (!req.body._id || !req.body.parent)
     {
-      var addedRole = new Roles();
-      addedRole._id = role._id;
-      addedRole.parent = role.parent || null;
-      role.canModify = role.canModify || false;
-      role.parentForDescendants = role.parentForDescendants || [];
-
-      //check to make sure the parent exists
-      if(addedRole.parent === null || !roleExists(addedRole.parent))
-      {
-        res.status(500).send("Parent must be a valid role");
-        return;
-      }
-
-
-      //save will fail if _id isn't unique
-      addedRole.save((err, data) => {
-        if(err)
+      return sendServerError(res, "required role fields not set");
+    }
+    //map request body params to a role
+    var role = new Roles();
+    role._id = req.body._id;
+    role.parent = req.body.parent;
+    role.canModify = req.body.canModify || false;
+    role.parentForDescendants = req.body.parentForDescendants || [];
+
+  
+    Roles.count({_id: req.body.parent}).exec()
+    .then( count =>
+    {
+        if(count < 1)
         {
-          res.status(500).send("There is already a role with that name");
-          return;
+          return new Promise((resolve, reject) =>
```

Comment:
You're actually mixing promises.  These are built in promises, but you have the q framework loaded.  I think you could do the same thing (even keeping the code the same) if you start the function with `var Promise = q.promise;`.  That way you will stay in the q promise structure.  This could potentially cause some pretty obfuscated errors later (I'm not even 100% sure what happens when you mix and match promise frameworks).